### PR TITLE
MRC-258_completed immunization resource schema

### PR DESCRIPTION
**Pull request by qabusalameh**

### CODE REVIEW

File: spark_auto_mapper_fhir/backbone_elements/immunization_performer_backbone_element.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    ImmunizationFunctionCode,
+)
+
+
+class ImmunizationPerformerBackboneElement(FhirBackboneElementBase):
+    def __init__(
+        self,
+        *,
+        id_: Optional[FhirId] = None,
+        function: Optional[FhirList[ImmunizationFunctionCode]] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/backbone_elements/immunization_performer_backbone_element.py:22]: missing CodeableConcept between FhirList and type code

File: spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class ImmunizationProtocolAppliedBackboneElement(FhirBackboneElementBase):
+    def __init__(
+        self,
+        *,
+        id_: Optional[FhirId] = None,
+        series: Optional[FhirString] = None,
+        authority: Optional[Reference[Organization]] = None,
+        targetDisease: Optional[CodeableConcept[ImmunizationTargetDiseaseCode]] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py:24]: missing FhirList between optional and codeable concept

File: spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        doseNumberPositiveInt: FhirPositiveInt,
+        doseNumberString: FhirString,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py:26]: you may need to add "= None" at the end of these; I know doesNumber is a mandatory field, but the output can't have both

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 4]
```diff
@@ -1,30 +1,172 @@
...
         meta: Optional[Meta] = None,
         extension: Optional[FhirList[Extension]] = None,
+        identifier: Optional[FhirList[Identifier]] = None,
+        status: ImmunizationSiteCode,
+        statusReason: Optional[CodeableConcept[ImmunizationStatusReasonCode]] = None,
+        vaccineCode: CodeableConcept[VaccineCode],
+        patient: Reference[Patient],
+        encounter: Optional[Reference[Encounter]] = None,
+        occurrenceDateTime: FhirDateTime,
+        occurrenceString: FhirString,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:75]: same as doseNumber, both can't be mandatory

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 5]
```diff
@@ -1,30 +1,172 @@
...
+        primarySource: Optional[FhirBoolean] = None,
+        reportOrigin: Optional[CodeableConcept[ImmunizationOriginCode]] = None,
+        location: Optional[Reference[Location]] = None,
+        manufacturer: Optional[Reference[Organization]] = None,
+        lotNumber: Optional[FhirString] = None,
+        expirationDate: Optional[FhirDate] = None,
+        site: Optional[CodeableConcept[ImmunizationSiteCode]] = None,
+        route: Optional[CodeableConcept[ImmunizationRouteCode]] = None,
+        doseQuantity: Optional[SimpleQuantity] = None,
+        performer: Optional[FhirList[ImmunizationFunctionCode]] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:86]: 'performer' needs to be split out as a backbone element definition

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 6]
```diff
@@ -1,30 +1,172 @@
...
+        note: Optional[Annotation] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:87]: FhirList

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 7]
```diff
@@ -1,30 +1,172 @@
...
+        reasonCode: Optional[CodeableConcept[ImmunizationReasonCode]] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:88]: FhirList

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 8]
```diff
@@ -1,30 +1,172 @@
...
+        reasonReference: Optional[
+            Reference[Union[Condition, Observation, "DiagnosticReport"]]
+        ] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:91]: FhirList

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 9]
```diff
@@ -1,30 +1,172 @@
...
+        subpotentReason: Optional[
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:92]: missing isSubpotent: Optinal[FhirBoolean] = None,

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 10]
```diff
@@ -1,30 +1,172 @@
...
+            CodeableConcept[ImmunizationSubpotentReasonCode]
+        ] = None,
+        education: Optional[ImmunizationEducationBackboneElement] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:95]: FhirList

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 11]
```diff
@@ -1,30 +1,172 @@
...
+        programEligibility: Optional[
+            CodeableConcept[ImmunizationProgramElegibilityCode]
+        ] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:98]: FhirList

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 12]
```diff
@@ -1,30 +1,172 @@
...
+        fundingSource: Optional[CodeableConcept[ImmunizationFundingSourceCode]] = None,
+        reaction: Optional[ImmunizationReactionBackboneElement] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:100]: FhirList

File: spark_auto_mapper_fhir/resources/immunization.py [hunk 13]
```diff
@@ -1,30 +1,172 @@
...
+        protocolApplied: Optional[ImmunizationTargetDiseaseCode] = None,
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/resources/immunization.py:101]: protocolApplied should be a backbone element

File: spark_auto_mapper_fhir/valuesets/immunization_function.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationFunctionCode("AP")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-function
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-function "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_function.py:53]: remove space

File: spark_auto_mapper_fhir/valuesets/immunization_funding_source.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationFundingSourceCode("public")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-funding-source
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-funding-source "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_funding_source.py:53]: remove space

File: spark_auto_mapper_fhir/valuesets/immunization_origin.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationOriginCode("school")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-origin
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-origin "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_origin.py:69]: space

File: spark_auto_mapper_fhir/valuesets/immunization_program_elegibility.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationProgramElegibilityCode("uninsured")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-program-eligibility
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-program-eligibility "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_program_elegibility.py:55]: space

File: spark_auto_mapper_fhir/valuesets/immunization_reason_code.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationReasonCode("A")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-reason
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-reason "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_reason_code.py:45]: space

File: spark_auto_mapper_fhir/valuesets/immunization_route.py [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationRouteCode("TRNSDERM")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-route
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-route "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_route.py:93]: space

File: spark_auto_mapper_fhir/valuesets/immunization_site.py [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationSiteCode("RA")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-site
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-site "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_site.py:53]: space

File: spark_auto_mapper_fhir/valuesets/immunization_status.py [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationStatusCode("not-done")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-status
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-status "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_status.py:64]: space

File: spark_auto_mapper_fhir/valuesets/immunization_status_reason.py [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationStatusReasonCode("A")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-status-reason
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-status-reason "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_status_reason.py:45]: space

File: spark_auto_mapper_fhir/valuesets/immunization_subpotent_reason.py [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationSubpotentReasonCode("recall")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-subpotent-reason
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-subpotent-reason "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_subpotent_reason.py:61]: space

File: spark_auto_mapper_fhir/valuesets/immunization_target_disease.py [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        # noinspection PyCallingNonCallable
+        return ImmunizationTargetDiseaseCode("A")
+
+    @genericclassproperty
+    def codeset(cls) -> FhirUri:
+        """
+        http://hl7.org/fhir/ValueSet/immunization-target-disease
+        """
+        return "http://hl7.org/fhir/ValueSet/immunization-target-disease "
```
Comment (abbey-icanbwell) [inline spark_auto_mapper_fhir/valuesets/immunization_target_disease.py:45]: space

**Review by abbey-icanbwell — CHANGES_REQUESTED** [review]

File: spark_auto_mapper_fhir/backbone_elements/immunization_performer_backbone_element.py (hunk 1 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/backbone_elements/immunization_performer_backbone_element.py:22]: Corrected

File: spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py (hunk 2 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py:24]: Done

File: spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py (hunk 3 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/backbone_elements/immunization_protocolApplied_backbone_element.py:26]: Thank you. Added the "=None,"

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 4 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:75]: Done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 5 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:86]: I created the BackboneElement but I didnt add, thank you for noticing it. I corrected it

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 7 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:88]: done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 8 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:91]: done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 9 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:92]: Done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 10 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:95]: done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 11 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:98]: done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 12 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:100]: done

File: spark_auto_mapper_fhir/resources/immunization.py (hunk 13 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/resources/immunization.py:101]: done

File: spark_auto_mapper_fhir/valuesets/immunization_function.py (hunk 14 above)
Comment (qabusalameh) [reply spark_auto_mapper_fhir/valuesets/immunization_function.py:53]: done

**Review by imranq2 — APPROVED** [review]

**Review by imranq2 — APPROVED** [review]

**Pull request merged**